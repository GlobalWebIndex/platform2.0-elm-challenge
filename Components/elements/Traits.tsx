import { TraitProps } from "interfaces/elements/Trait";
import useTranslation from "next-translate/useTranslation";

export const Traits = (props: TraitProps) => {
    const keys = Object.keys(props);
    const values = Object.values(props);
    const { t } = useTranslation("details");

    return (
        <section className="border-b-[1px] pb-4 md:pb-8 border-b-gray-300 dark:border-b-gray-800 mb-8 grid sm:grid-cols-3 lg:grid-cols-3 gap-y-2 gap-x-10">
            {
                keys.map((trait, index: number) => {
                    return (
                        <div key={trait + index} className="flex items-center">
                            <span className="text-gray-400 w-1/3">{t(trait)}</span>
                            <span className="text-gray-400 w-2/3 flex">
                                <span style={{width: `${values[index] * 20}%`}} className={`inline-block h-[20px] bg-gray-200 dark:bg-gray-400 backdrop-blur-lg`} />
                            </span>
                        </div>
                    )
                })
            }
        </section>
    )
};
