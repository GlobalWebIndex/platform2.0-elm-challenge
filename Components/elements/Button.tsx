import { ButtonProps } from "interfaces/elements/Button";
import { forwardRef } from "react";
import Link from "next/link";

export const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
    const { link, type = "button", onClick = () => { }, className = "", disabled = false } = props;

    return (
        <Link href={link.href}>
            <button data-testid="button" ref={ref} {...{ disabled }} className={`disabled:opacity-0 disabled:pointer-events-none bg-gradient-to-br from-gray-200 to-gray-100 dark:from-gray-800 dark:to-gray-900 backdrop-blur-lg px-12 py-2 md:py-3 block ${className}`} type={type} {...{ onClick }}>
                {link.label}
            </button>
        </Link>
    )
});

Button.displayName = 'Button';