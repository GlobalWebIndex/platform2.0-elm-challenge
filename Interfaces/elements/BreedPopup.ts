export interface BreedPopupProps {
    name: undefined | null | string | string[],
    onClose?: () => void
};