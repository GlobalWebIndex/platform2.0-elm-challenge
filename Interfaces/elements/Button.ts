import { LinkProps } from "interfaces/general/Link";
import { MouseEvent } from "react";

export type ButtonTypes = "button" | "submit" | "reset" | undefined;

export interface ButtonProps {
    link: LinkProps,
    type?: ButtonTypes,
    className?: string,
    disabled?: boolean,
    onClick?: (arg: MouseEvent<HTMLButtonElement>) => void
};