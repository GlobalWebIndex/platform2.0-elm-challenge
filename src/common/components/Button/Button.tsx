import styled from 'styled-components';
import { Colors } from 'theme';

const SButton = styled.button<{ variant?: string }>`
  width: 100%;
  height: 100%;
  background: ${({ variant }) =>
    variant === 'primary'
      ? Colors.primary
      : variant === 'secondary'
      ? Colors.white
      : variant === 'danger'
      ? Colors.danger
      : Colors.primary};

  :hover {
    background: ${({ variant }) =>
      variant === 'primary'
        ? Colors.primaryHover
        : variant === 'secondary'
        ? Colors.secondaryHover
        : variant === 'danger'
        ? Colors.dangerHover
        : Colors.primary};
  }
`;

interface Props {
  text: string;
  disabled?: boolean;
  variant?: 'primary' | 'secondary' | 'danger';
  children?: React.ReactNode;
  onClick: () => void;
}

const Button = ({ variant, disabled = false, text, onClick }: Props) => {
  return (
    <SButton
      variant={variant}
      disabled={disabled}
      onClick={onClick}
      type="button"
    >
      {text}
    </SButton>
  );
};

export default Button;
