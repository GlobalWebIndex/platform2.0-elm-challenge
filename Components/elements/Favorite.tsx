import { HeartIcon as HeartFullIcon } from "@heroicons/react/24/solid";
import { AppContext } from "context/AppProvider";
import { ContextProps } from "interfaces/context/Context";
import { useContext, useState } from "react";
import { FavoriteProps } from "interfaces/elements/Favorite";
import { fetchData } from "helpers/net/fetchData";
import { endpoints } from "configuration/endpoints";
import { constants } from "configuration/constants";

export const Favorite = ({ imageId }: FavoriteProps) => {
    const [disabled, setDisabled] = useState(false);
    const { favorites, addToFavorites, removeFromFavorites } = useContext<ContextProps>(AppContext);
    const favorite = Array.isArray(favorites) && favorites.find((item) => item.imageId === imageId);

    //call the api service to add to favorites; if successful, update our internal state
    const handleAddToFavorites = async () => {
        const result = await fetchData({method: "post", data: {image_id: imageId, sub_id: constants.sub_id}, endpoint: endpoints.favorite, onStart: () => setDisabled(true), onEnd: () => setDisabled(false), apikey: constants.apikey});

        if(imageId && result?.message === "SUCCESS" && typeof addToFavorites === "function")
            addToFavorites({ imageId: imageId, favoriteId: result?.id });
    };

    //call the api service to remove from favorites; if successful, update our internal state
    const handleRemoveFromFavorites = async () => {
        const id = (favorite) && favorite?.favoriteId;
        const result = await fetchData({method: "delete", endpoint: `${endpoints.favorite}${id}`, onStart: () => setDisabled(true), onEnd: () => setDisabled(false), apikey: constants.apikey});

        if(imageId && id && result?.message === "SUCCESS" && typeof removeFromFavorites === "function")
            removeFromFavorites({ imageId: imageId, favoriteId: id });
    };

    return (
        <div className={`hover:scale-120 transition-transform absolute top-4 left-4 cursor-pointer z-10 ${(disabled) && "pointer-events-none scale-90"}`} onClick={() => {(!favorite) ? handleAddToFavorites() : handleRemoveFromFavorites() }}>
            <HeartFullIcon className={`drop-shadow-lg ${(!favorite) ? "text-white" : "text-red-600"}`} width={32} height={32} />
        </div>
    )
};