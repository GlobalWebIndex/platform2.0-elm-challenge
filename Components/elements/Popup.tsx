import { XMarkIcon } from "@heroicons/react/24/solid";
import { PopupProps } from "interfaces/elements/Popup";
import { createPortal } from "react-dom";

export const Popup = ({ onClose = () => {}, children }: PopupProps) => {
    const popupMarkup = (
        <div className="fixed top-0 left-0 w-screen h-screen z-20 flex justify-center items-center">
            <div className="w-full h-full fixed z-0 backdrop-blur-sm" onClick={onClose} />
            
            <div className="relative w-[90%] h-[85%] md:w-2/3 md:h-2/3 bg-white/90 dark:bg-black/90 backdrop-blur-lg z-10">
                <div className="absolute right-4 top-4 cursor-pointer z-10 bg-white/75 dark:bg-black/75 text-gray-700 dark:text-gray-400 backdrop-blur-lg">
                    <XMarkIcon className="hover:scale-90 transition-transform" onClick={onClose} width={48} height={48} />
                </div>
                {children}
            </div>
        </div>
    );

    return (createPortal(popupMarkup, document.getElementById("tailwind")!))
};