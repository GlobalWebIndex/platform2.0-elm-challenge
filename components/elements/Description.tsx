import { DescriptionProps } from "interfaces/elements/Description";
import useTranslation from "next-translate/useTranslation";

export const Description = ({ alt_names, description, life_span, temperament }: DescriptionProps) => {
    const { t } = useTranslation("breed");

    return (
        <>
            <p className="mb-8">{description}</p>

            {
                (alt_names) &&
                <div className="mb-4 text-sm text-gray-400">
                    <p>{t("alsoknownas")}</p>
                    <p>{alt_names}</p>
                </div>
            }
            {
                (temperament) &&
                <div className="mb-4 text-sm text-gray-400">
                    <p>{t("temperament")}</p>
                    <p>{temperament}</p>
                </div>
            }
            {
                (life_span) &&
                <div className="mb-4 text-sm text-gray-400">
                    <p>{t("lifespan")}</p>
                    <p>{life_span} {t("years")}</p>
                </div>
            }
        </>
    )
};