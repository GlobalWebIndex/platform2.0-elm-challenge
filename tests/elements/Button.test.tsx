import { render, screen } from '@testing-library/react';
import { Button } from 'components/elements/Button';
import '@testing-library/jest-dom';

const link = {
    label: "Press me",
    href: "/"
};

describe('Button', () => {
    it('Button is disabled when we pass the prop', () => {
        render(
            <Button {...{ link }} disabled />
        );

        expect(screen.getByTestId("button")).toHaveAttribute("disabled");
    });

    it("renders correctly", () => {
        const { container } = render(
            <Button {...{ link }} disabled />
        );
        expect(container).toMatchSnapshot();
    });
});