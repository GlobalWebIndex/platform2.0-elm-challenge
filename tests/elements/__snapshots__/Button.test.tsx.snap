// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div>
  <button
    class="disabled:opacity-0 disabled:pointer-events-none bg-gradient-to-br from-gray-200 to-gray-100 dark:from-gray-800 dark:to-gray-900 backdrop-blur-lg px-12 py-2 md:py-3 block "
    data-testid="button"
    type="button"
  >
    Press me
  </button>
</div>
`;
